components/director/internal/domain/formation/service.go

```diff
@@ -1002,106 +1062,124 @@ func (s *service) ResynchronizeFormationNotifications(ctx context.Context, forma
-		if err != nil && !apperrors.IsNotFoundError(err) {
-			return nil, err
-		}
-		if reverseFA != nil {
-			notificationForReverseFA, err = s.formationAssignmentNotificationService.GenerateFormationAssignmentNotification(faResyncTransactionCtx, reverseFA, operation)
+		for _, fa := range resyncableFormationAssignments {
+			operation := model.AssignFormation
+			if fa.State == string(model.DeleteErrorAssignmentState) || fa.State == string(model.DeletingAssignmentState) {
+				operation = model.UnassignFormation
+			}
+			var notificationForReverseFA *webhookclient.FormationAssignmentNotificationRequest
+			notificationForFA, err := s.formationAssignmentNotificationService.GenerateFormationAssignmentNotification(ctxWithTransact, fa, operation)
+			if err != nil {
+				return err
+			}
+
+			reverseFA, err := s.formationAssignmentService.GetReverseBySourceAndTarget(ctxWithTransact, fa.FormationID, fa.Source, fa.Target)
 			if err != nil && !apperrors.IsNotFoundError(err) {
-				return nil, err
+				return err
+			}
+			if reverseFA != nil {
+				notificationForReverseFA, err = s.formationAssignmentNotificationService.GenerateFormationAssignmentNotification(ctxWithTransact, reverseFA, operation)
+				if err != nil && !apperrors.IsNotFoundError(err) {
+					return err
+				}
 			}
-		}
 
-		var reverseReqMapping *formationassignment.FormationAssignmentRequestMapping
-		if reverseFA != nil || notificationForReverseFA != nil {
-			reverseReqMapping = &formationassignment.FormationAssignmentRequestMapping{
-				Request:             notificationForReverseFA,
-				FormationAssignment: reverseFA,
+			var reverseReqMapping *formationassignment.FormationAssignmentRequestMapping
+			if reverseFA != nil || notificationForReverseFA != nil {
+				reverseReqMapping = &formationassignment.FormationAssignmentRequestMapping{
+					Request:             notificationForReverseFA,
+					FormationAssignment: reverseFA,
+				}
 			}
-		}
 
-		assignmentPair := formationassignment.AssignmentMappingPairWithOperation{
-			AssignmentMappingPair: &formationassignment.AssignmentMappingPair{
-				AssignmentReqMapping: &formationassignment.FormationAssignmentRequestMapping{
-					Request:             notificationForFA,
-					FormationAssignment: fa,
-				},
-				ReverseAssignmentReqMapping: reverseReqMapping,
-			},
-		}
-		switch fa.State {
-		case string(model.InitialAssignmentState), string(model.CreateErrorAssignmentState):
-			assignmentPair.Operation = model.AssignFormation
-			if _, err = s.formationAssignmentService.ProcessFormationAssignmentPair(faResyncTransactionCtx, &assignmentPair); err != nil {
-				errs = multierror.Append(errs, err)
+			if notificationForFA != nil && notificationForFA.Webhook.Mode != nil && *notificationForFA.Webhook.Mode == graphql.WebhookModeAsyncCallback && operation == model.UnassignFormation {
+				fa.State = string(model.DeletingAssignmentState)
```

Comment:
IMHO I think it's better to be handled in the `ProcessFormationAssignmentPair` and not in the resync method because that way it's encapsulated in the `ProcessFormationAssignmentPair ` method and if someone is using it won't have to prepare the assignments in INITIAL state, he will directly call the method. E.g. if we do it in the resync method, we have to do it for all other places using the `ProcessFormationAssignmentPair ` method - `formationmapping` package(`processFormationAssignmentNotifications`), assign/ProcessFormationAssignments in the `formationassignment` package, and potentially other places.

We are doing it for Unassign because the management of the "DELETING" state is no longer handled in the `CleanupFormationAssignment` method but rather in the `UnassignFormation` method. And `CleanupFormationAssignment` is used both in `UnassignFormation` and `resync` methods. That's why we have to handle this "async callback deleting" state in the resync method as well.

Reply:
The assignments should not be set to INITIAL in the ProcessFormationAssignmentPair. Setting it there is a workaround as the opertion is deduced from the state. Also if we keep setting the state to initial in ProcessFormationAssignmentPair we will have to introduce another transaction there as this is the reason for the rece conditions in assign

Reply:
Discussed offline and it's a separate issue in the Assign method that will be fix as part of another PR.